src/xopen/__init__.py

```diff
@@ -402,6 +402,24 @@ def xopen(filename, mode='r', compresslevel=6, threads=None):
         return _open_xz(filename, mode)
     elif filename.endswith('.gz'):
         return _open_gz(filename, mode, compresslevel, threads)
+    elif mode.startswith("r"):
+        # Test up to the first 6 bytes to detect the file format.
+        with open(filename, "rb") as fh:
```

Comment:
There are some circumstances under which this can fail, for example, if the given filename is `/dev/stdin` or `/dev/fd/0`. That would currently work as expected because there’s only one call to `open()`. However, with this change, there would be two calls and then the second would fail.

Can you please change it such that the file is only opened and read for content detection when it is a regular file? You could check this with `stat.S_ISREG(os.stat(filename).st_mode)`.

Another [solution for a similar file-detection problem, which I implemented in another project](https://github.com/marcelm/dnaio/blob/24f94975f2ac09f682a5486cc08fda4e6727fbcc/src/dnaio/__init__.py#L197) is to open the file just once, read from it, seek back to the beginning and then pass on the already open file descriptor (or use `peek()`). Here, this would imply that, for example, `gzip.open` would need to get an open, file-like object (this is possible with `gzip.open(fileobj=...)`), but then when you call `.close()` on the returned GzipFile, the underlying file descriptor is not closed. So while this method would even allow you to correctly open gzip-compressed standard input, there would be some bigger changes involved.

Reply:
This is a good point I hadn't considered. I added the check you suggested: `stat.S_ISREG(os.stat(filename).st_mode)`. However, if the file doesn't exist that returns an `OSError`. At the moment I change this to an `IOError` but not sure that is the best approach.